### PR TITLE
Implement OAuth Write Popup C-2664

### PR DESCRIPTION
**Pull request by nicoback2**
### Description

Designs: https://www.figma.com/file/52mX38f4QQkHgzUFRxbUnD/%E2%9A%99%EF%B8%8F-OAuth-%2B-Write-SDK-UI?type=design&node-id=150-2697&t=3qiqfUSypUx0yOgh-0

Not signed in:
Screenshot 2023-06-20 at 8 19 34 PM

Signed in, already authorized app:
Screenshot 2023-06-20 at 10 25 05 PM

Signed in, hasn't authorized app:
Screenshot 2023-06-20 at 10 26 20 PM

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

### How Has This Been Tested?

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### Feature Flags ###

*Are all new features properly feature flagged? Describe added feature flags.*

### CODE REVIEW

**Comment by audius-infra** [comment]
It looks like there may be some changes to native mobile code, which requires triggering a full app release. Please follow the instructions here: https://www.notion.so/audiusproject/When-to-bump-app-version-2644a8f772364a4d91f44abcba44ce0b?pvs=4. cc @nicoback2 @sliptype

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/nkang--oauth-popup-write

File: packages/web/src/pages/oauth-login-page/utils.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
+export const getIsRedirectValid = ({
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/utils.ts:1]: love how you added this utils page

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 2]
```diff
@@ -52,15 +76,50 @@ const messages = {
...
+  })
+}
+
+const getDeveloperApp = async (address: string) => {
+  const sdk = await audiusSdk()
+  const developerApp = await sdk.developerApps.getDeveloperApp({ address })
+  return developerApp.data
+}
+
+const getIsAppAuthorized = async ({
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:106]: would we want this to be in utils file also?

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 3]
```diff
@@ -1,19 +1,31 @@
-import { FormEvent, useEffect, useLayoutEffect, useMemo, useState } from 'react'
+import {
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:1]: This file is getting pretty large, all good if we just want it one big page, but i wonder if a bit more modularity might help with long-term readability and feature updates?

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 4]
```diff
@@ -83,6 +142,112 @@ const CTAButton = ({
...
+            ) : (
+              <IconVisibilityPublic
+                className={cn(styles.permissionIcon, styles.visibilityIcon)}
+                width={21}
+                height={22}
+              />
+            )}
+          </div>
+
+          <div className={styles.permissionTextContainer}>
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:183]: p tag?

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 5]
```diff
@@ -83,6 +142,112 @@ const CTAButton = ({
...
+                  <br />
+                  {messages.walletsOrDMs}
+                </p>
+              </div>
+            )}
+          </div>
+        </div>
+        <div
+          className={cn(
+            styles.permissionContainer,
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:207]: fine like this, but you can also use the `composes: permissionContainer` in the style block for `nonFirstPermissionContainer` in the css modules file. lmk if interested in that pattern.

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 6]
```diff
@@ -83,6 +142,112 @@ const CTAButton = ({
...
+                <span
+                  className={cn(
+                    styles.permissionText,
+                    styles.permissionDetailText,
+                    {
+                      [styles.permissionTextExtraLight]: !userEmail
+                    }
+                  )}
+                >
+                  {userEmail == null ? (
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:233]: ===

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 7]
```diff
@@ -83,6 +142,112 @@ const CTAButton = ({
...
+                    <>
+                      <LoadingSpinner className={styles.loadingSpinner} /> Email
+                      loading...
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:236]: should we add to messages object? also, can you use html for the ellipses? not all screenreaders do this correctly: https://www.htmlsymbol.com/html-entities/horizontal-ellipsis.html

Comment (dylanjeffers) [reply packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:236]: checkout PublishButton for example

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 8]
```diff
@@ -119,27 +282,76 @@ export const OAuthLoginPage = () => {
...
   }, [redirect_uri])
-  const account = useSelector(getAccountUser)
-  const isLoggedIn = Boolean(account)
-  const [userEmail, setUserEmail] = useState<string | null>(null)
-  useEffect(() => {
-    const getAndSetEmail = async () => {
-      let email: string
+
+  const isRedirectValid = useMemo(() => {
+    return getIsRedirectValid({ parsedRedirectUri, redirectUri: redirect_uri })
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:287]: just double checking this is just to avoid computation rather than avoiding rerenders?

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 9]
```diff
@@ -119,27 +282,76 @@ export const OAuthLoginPage = () => {
...
-      setUserEmail(email)
     }
-    if (isLoggedIn) {
-      getAndSetEmail()
-    } else {
-      setUserEmail(null)
+    return null
+  }, [originParam])
+
+  let initError: string | null = null
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:301]: def a minor refactor  thing, would would be just making a "getError" function or something rather?

**Comment by audius-infra** [comment]
It looks like there may be some changes to native mobile code, which requires triggering a full app release. Please follow the instructions here: https://www.notion.so/audiusproject/When-to-bump-app-version-2644a8f772364a4d91f44abcba44ce0b?pvs=4. cc @nicoback2 @sliptype

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 10]
```diff
@@ -263,104 +410,226 @@ export const OAuthLoginPage = () => {
     response_mode
   ])
 
-  const formOAuthResponse = async (account: User) => {
-    let email: string
-    if (!userEmail) {
+  useEffect(() => {
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:413]: would we want a "getDeveloperApp" method in developer-app api?

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx [hunk 11]
```diff
@@ -411,14 +684,23 @@ export const OAuthLoginPage = () => {
       setAndLogGeneralSubmitError(false, messages.miscError)
     } else {
       setIsSubmitting(true)
-      authAndRedirect(account)
+      authorize(account)
```
Comment (dylanjeffers) [inline packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:687]: do we want to await here?

**Review by dylanjeffers — APPROVED** [review]
TONS of great changes here. I would say high level its a bit hard to follow all the logic here, and wonder if there is a way to bucket general flows into their own files/hooks etc? Up to you on the lift to do that is worth it though. would also be sick to see the developer-app api get another method or two for this page.

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/nkang--oauth-popup-write

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx (hunk 3 above)
Comment (nicoback2) [reply packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:1]: yeah i had the same thought, will put some thought into this

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx (hunk 7 above)
Comment (nicoback2) [reply packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:236]: ah yes! thanks for catching

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx (hunk 8 above)
Comment (nicoback2) [reply packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:287]: yeah!

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx (hunk 10 above)
Comment (nicoback2) [reply packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:413]: thought about this but decided against since we don't need the result to be cached or anything! but can play around and see if it makes the code simpler

File: packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx (hunk 11 above)
Comment (nicoback2) [reply packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx:687]: we don't need to because nothing comes after

**Comment by audius-infra** [comment]
It looks like there may be some changes to native mobile code, which requires triggering a full app release. Please follow the instructions here: https://www.notion.so/audiusproject/When-to-bump-app-version-2644a8f772364a4d91f44abcba44ce0b?pvs=4. cc @nicoback2 @sliptype

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/nkang--oauth-popup-write
